test/e2e/provider_expectations.go

```diff
@@ -298,3 +306,316 @@ func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
+	// random node and observe reconciliation of pods on the other one.
+	if err := tc.client.Delete(context.TODO(), machine1); err != nil {
+		return fmt.Errorf("unable to delete machine %q: %v", machine1.Name, err)
+	}
+
+	delete(delObjects, "machine1")
+
+	// We still should be able to list the machine as until rc.replicas-1 are running on the other node
+	if err := wait.PollImmediate(1*time.Second, 10*time.Minute, func() (bool, error) {
+		machine := mapiv1beta1.Machine{}
+
+		key := types.NamespacedName{
+			Namespace: machine1.Namespace,
+			Name:      machine1.Name,
+		}
+		if err := tc.client.Get(context.TODO(), key, &machine); err != nil {
+			glog.Errorf("error querying api machine %q object: %v, retrying...", machine1.Name, err)
+			return false, nil
+		}
+		if machine.Status.NodeRef == nil || machine.Status.NodeRef.Kind != "Node" {
+			glog.Error("machine %q not linked to a node", machine.Name)
+		}
+
+		pods := corev1.PodList{}
+		listOpt := &client.ListOptions{}
+		listOpt.MatchingLabels(rc.Spec.Selector)
+		if err := tc.client.List(context.TODO(), listOpt, &pods); err != nil {
+			glog.Errorf("error querying api for Pods object: %v, retrying...", err)
+			return false, nil
+		}
+
+		// expecting nodeGroupSize nodes
+		podCounter := 0
+		for _, pod := range pods.Items {
+			if pod.Spec.NodeName != machine.Status.NodeRef.Name {
+				continue
+			}
+			if !pod.DeletionTimestamp.IsZero() {
+				continue
+			}
+			podCounter++
+		}
+
+		glog.Infof("Have %v pods scheduled to node %q", podCounter, machine.Status.NodeRef.Name)
+
+		// Verify we have enough pods running as well
+		rcObj := corev1.ReplicationController{}
+		key = types.NamespacedName{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+		}
+		if err := tc.client.Get(context.TODO(), key, &rcObj); err != nil {
+			glog.Errorf("error querying api RC %q object: %v, retrying...", rc.Name, err)
+			return false, nil
+		}
+
+		// The point of the test is to make sure majority of the pods is rescheduled
+		// to other nodes. Pod disruption budget makes sure at most one pod
+		// owned by the RC is not Ready. So no need to test it. Though, usefull to have it printed.
+		glog.Infof("RC ReadyReplicas/Replicas: %v/%v", rcObj.Status.ReadyReplicas, rcObj.Status.Replicas)
+
+		// This makes sure at most one replica is not ready
+		if rcObj.Status.Replicas-rcObj.Status.ReadyReplicas > 1 {
+			return false, fmt.Errorf("pod disruption budget not respecpted, node was not properly drained")
+		}
+
+		// Depends on timing though a machine can be deleted even before there is only
+		// one pod left on the node (that is being evicted).
+		if podCounter > 2 {
+			glog.Infof("Expecting at most 2 pods to be scheduled to drained node %v, got %v", machine.Status.NodeRef.Name, podCounter)
+			return false, nil
+		}
+
+		glog.Info("Expected result: all pods from the RC up to last one or two got scheduled to a different node while respecting PDB")
+		return true, nil
+	}); err != nil {
+		return err
```

Comment:
we are we validating the machine object and the backed node does not exist anymore?

Reply:
if the machine object does not get removed, the loop above fails due to
```go
if err := tc.client.Get(context.TODO(), key, &machine); err != nil {
	glog.Errorf("error querying api machine %q object: %v, retrying...", machine1.Name, err)
	return false, nil
}
```

The backed up node is actually removed later (by cloud controller manager once it goes NotReady). It can happen in 1 minute, in 2 minutes, later, sooner? So, we can't verify that.

Reply:
yes we can, and we should. We are actually validating that in other test cases

Reply:
We can do it though this test is about verifying a node is drained before machine object is deleted. It does not test a node is deleted. Since testing "node is deleted after linked machine is delete" is another test case.

Reply:
This should test e2e features from a product end user pov. This test is pretty much validating "Pod disruption budget" and eviction which is only a part of the user expected e2e feature. Then it is assuming that the pods were moved because the machine was deleted (but that could be untrue. Correct draining could be a result of buggy code on a controller telling to drain but then failing to delete the machine and as a result the node).
We want to validate in code explicitly exactly the end user story: When I delete a machine the node is drained (covered by the test), then machine is deleted (not covered), then node is delete  (not covered)